replay2/memory_remapper/tests/memory_remapper_tests.cc

```diff
@@ -0,0 +1,205 @@
+		}
+	}
+private:
+	std::byte value_;
+	size_t length_;
+};
+
+void ASSERT_MOD_REPLAY_ADDRESS(const MemoryRemapper& remapper, const CaptureAddress& captureAddress, const ReplayAddress& replayAddress, size_t length) {
+	for(size_t i = 0; i < length; ++i) {
+		ReplayAddress replayAddress = remapper.RemapCaptureAddress(captureAddress.offsetByBytes(i));
+		ASSERT_EQ(replayAddress.bytePtr()[0], std::byte(i % 256));
+	}
+}
+
+void ASSERT_CONST_REPLAY_ADDRESS(const MemoryRemapper& remapper, const CaptureAddress& captureAddress, const ReplayAddress& replayAddress, std::byte value, size_t length) {
+	for(size_t i = 0; i < length; ++i) {
+		ReplayAddress replayAddress = remapper.RemapCaptureAddress(captureAddress.offsetByBytes(i));
+		ASSERT_EQ(replayAddress.bytePtr()[0], value);
+	}
+}
+
+TEST(MemoryRemapperTests, SimpleMapping) {
+
+	const size_t size = 128;
+
+	std::byte* rawCapturePtr = new std::byte[size];
+	CaptureAddress captureAddress(rawCapturePtr);
+
+	MemoryRemapper remapper;
+	const MemoryObservation captureObservation(captureAddress, std::make_shared<ModResourceGenerator>(size));
+
+	const ReplayAddress replayAddress = remapper.AddMapping(captureObservation);
+	EXPECT_NE(replayAddress.bytePtr(), nullptr);
+	ASSERT_MOD_REPLAY_ADDRESS(remapper, captureAddress, replayAddress, size);
+
+	const ReplayAddress replayAddress2 = remapper.RemapCaptureAddress(captureAddress);
+	EXPECT_EQ(replayAddress, replayAddress2);
+
+	EXPECT_NO_THROW(remapper.RemoveMapping(captureAddress));
+	EXPECT_THROW(remapper.RemapCaptureAddress(captureAddress), MemoryRemapper::AddressNotMappedException);
+}
+
+TEST(MemoryRemapperTests, UnknownMapping) {
+
+	const size_t size = 128;
+
+	std::byte* rawCapturePtr = new std::byte[size];
+	CaptureAddress captureAddress(rawCapturePtr);
+
+	MemoryRemapper remapper;
+
+	EXPECT_THROW(remapper.RemapCaptureAddress(captureAddress), MemoryRemapper::AddressNotMappedException);
+
+	EXPECT_THROW(remapper.RemoveMapping(captureAddress), MemoryRemapper::AddressNotMappedException);
+	EXPECT_THROW(remapper.RemapCaptureAddress(captureAddress), MemoryRemapper::AddressNotMappedException);
+}
+
+TEST(MemoryRemapperTests, ZeroLengthMapping) {
+
+	const size_t size = 0;
+
+	std::byte* rawCapturePtr = new std::byte[size];
+	CaptureAddress captureAddress(rawCapturePtr);
+
+	MemoryRemapper remapper;
+	const MemoryObservation captureObservation(captureAddress, std::make_shared<ModResourceGenerator>(size));
+
+	EXPECT_THROW(remapper.AddMapping(captureObservation), MemoryRemapper::CannotMapZeroLengthAddressRange);
+	EXPECT_THROW(remapper.RemapCaptureAddress(captureAddress), MemoryRemapper::AddressNotMappedException);
+	EXPECT_THROW(remapper.RemoveMapping(captureAddress), MemoryRemapper::AddressNotMappedException);
+}
+
+TEST(MemoryRemapperTests, MultipleMappings) {
+
+	std::vector<ReplayAddress> replayAddresses;
+	std::vector<CaptureAddress> captureAddresses;
+
+	MemoryRemapper remapper;
+
+	for(int i = 0; i < 64; ++i) {
```

Comment:
These are tests but it would still be helpful to have names.

Reply:
Names for what? The variable i?